### PR TITLE
feat(querying): Add declarative querying

### PR DESCRIPTION
**Pull request by davideast**
@jeffbcross Ready for your review!

**Changelog:**
- Added interactive querying! #16 
- Emit errors from security rules. #146 

Examples:

With Scalar values

```ts
class MyComponent {
  constructor(af: AngularFire) {
    const subject = new Subject();
    const peopleStream = af.database.list('/people', {
      query: {
        orderByChild: 'height',
        startAt: 60
      }
    });
    peopleStream.subscribe(people => console.log(people);
  }
}
```

With Observables 😱 😱 😱

```ts
class MyComponent {
  constructor(af: AngularFire) {
    const subject = new Subject();
    const peopleStream = af.database.list('/people', {
      query: {
        orderByChild: 'height',
        startAt: subject
      }
    });
    peopleStream.subscribe(people => console.log(people);
    subject.next(60); // re-run the query!
  }
}
```

Emitting error:

```ts
    const queryObs = af.database.list('/messages').subscribe(
      data => console.log(data),
      err => console.error(err)
    );
```

### CODE REVIEW

File: src/utils/firebase_list_factory.ts [hunk 1]
```diff
@@ -2,25 +2,90 @@ import {FirebaseListObservable, AFUnwrappedDataSnapshot} from './firebase_list_o
...
+       utils.isEmptyObject(query)) {
+    return firebaseListObservable(ref, { preserveSnapshot });
+  }
   
-  return new FirebaseListObservable((obs:Observer<any[]>) => {
-    let arr:any[] = [];
-    let hasInitialLoad = false;
+  const queryObs = observeQuery(query);
+  const listObs = <FirebaseListObservable<{}>>queryObs
+    .map(queryOrder => {
```
Comment (jeffbcross) [inline src/utils/firebase_list_factory.ts]: Should be renamed something more generic like "query"

File: src/utils/firebase_list_factory.ts [hunk 2]
```diff
@@ -2,25 +2,90 @@ import {FirebaseListObservable, AFUnwrappedDataSnapshot} from './firebase_list_o
...
+        
+        // apply limitTos
+        if (utils.isPresent(queryOrder.limitToFirst)) {
+          queried = queried.limitToFirst(queryOrder.limitToFirst);
+        }
+        
+        if (utils.isPresent(queryOrder.limitToLast)) {
+          queried = queried.limitToLast(queryOrder.limitToLast);
+        }
+
```
Comment (jeffbcross) [inline src/utils/firebase_list_factory.ts]: Raise an exception if they also have startAt or endAt

File: src/utils/firebase_list_factory.ts [hunk 3]
```diff
@@ -2,25 +2,90 @@ import {FirebaseListObservable, AFUnwrappedDataSnapshot} from './firebase_list_o
...
+      if (utils.isPresent(queryOrder.startAt)) {
+          queried = queried.startAt(queryOrder.startAt);
+      }
+      
+      if (utils.isPresent(queryOrder.endAt)) {
+          queried = queried.endAt(queryOrder.endAt);
+      }
+      
+      // apply limitTos
+      if (utils.isPresent(queryOrder.limitToFirst)) {
```
Comment (jeffbcross) [inline src/utils/firebase_list_factory.ts]: Raise an exception if multiple limitTos are used.

File: src/utils/query_observable.spec.ts [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+          orderByPriority: true,
+          endAt: new Subject(),
+          startAt: new Subject()
+        };
+        observableQueryTest(query, { startAt: 7, endAt: 12 }, done);
+      });            
+                        
+    });
+       
+});
```
Comment (jeffbcross) [inline src/utils/query_observable.spec.ts]: Newline plz

**Comment by jeffbcross** [comment]
A couple of comments to address, and needs some docs.

File: src/utils/query_observable.spec.ts (hunk 4 above)
Comment (davideast) [inline src/utils/query_observable.spec.ts]: Not sure what's going on, but my newlines are being ignored.

**Comment by davideast** [comment]
Will add docs in a separate MEGA-DOC PR.

**Pull request merged**